chore: named callbacks ADR (#676)
defenseunicorns/pepr: adr/0008-named-callbacks.md

> # 7: LogAs Function in Capability Class "Named Callbacks"

Reviewer: "LogAs" seems kinda limited to me. :thinking: 

We talked about this being a general purpose feature that could / would (essentially) carry a label for any given callback around throughout the Pepr system, right?  As it stands, the only place we're currently showing callback function names is in the log messages but that's not all we think can / will be done with this IMO, and I'd like to try to avoid handicapping ourselves by giving it an overly limiting name.

What would you say to naming this something that hinted at a more general usage?  Something like:
- Alias()
- Designation()
- Handle()
- or maybe simply, As()

All of those give a good sense of what we want this new func to be thought of as _without_ necessarily limiting it to logs.

Author: Interesting prospect here, not opposed by any means. Where else other than logs do you see this name being used?

Author: We like Alias because a user who has not read the docs is most likely to interpret this as a name.

- Pepr Sync

> # 7: LogAs Function in Capability Class "Named Callbacks"

Author: ```suggestion
# 7: Alias Function in Capability Class "Named Callbacks"
```

Date: 2024-03-21

## Status

> Proposed

Author: ```suggestion
Delayed - Revisiting this in the future and recommending that named callbacks be used in the interim. 
```

> Delayed - Revisiting this in the future and recommending that named callbacks be used in the interim. 

Reviewer 2: Accepted

> Delayed - Revisiting this in the future and recommending that named callbacks be used in the interim. 

Author: ```suggestion
Accepted.
```

## Context

> Pepr utilizes anonymous callbacks that produce logs during execution. This approach, while functional, has led to challenges in debugging and testing. The logs generated by these callbacks lack clear identifiers, making it difficult to ascertain which callback produced which log unless manual names are embedded in the log messages from within the callbacks themselves.

Reviewer: ```suggestion
Currently, Pepr utilizes function callback names to produce logs during execution. This approach -- when configured via nameless arrow function expression -- leads to challenges in debugging and testing. The logs generated by these callbacks lack clear, module author defined identifiers, making it difficult to ascertain which callback produced which effect unless manual names are embedded into log messages from within the callbacks themselves.
```

## Decision Drivers

The key drivers for this decision include:

> - **Devex**: Instead of manual declaration of `const name = "my-action";` in each callback and each log, we can add a name to the callback to make it easier to identify the source of logs. `LogAs("my-action")`

Reviewer: ```suggestion
- **Devex**: Instead of manual declaration of `const name = "my-action";` in each callback and each log (which is easy to forget and tedious to do), we can add a name to the callback to make it easier to identify the source of logs. `LogAs("my-action")`
```

> - **Devex**: Instead of manual declaration of `const name = "my-action";` in each callback and each log (which is easy to forget and tedious to do), we can add a name to the callback to make it easier to identify the source of logs. `LogAs("my-action")`

Author: ```suggestion
- **Devex**: Instead of manual declaration of `const name = "my-action";` in each callback and each log (which is easy to forget and tedious to do), we can add a name to the callback to make it easier to identify the source of logs. `Alias("my-action")`
```

- **Debugging Efficiency**: The debugging process in the current system is time-consuming and error-prone due to the anonymous nature of callbacks.  
- **Code Readability**: Anonymous callbacks can make the code harder to understand, especially for new developers or when returning to old code that has a different form of naming.
- **Maintenance and Scalability**: As the system grows, the manageability of anonymous callbacks becomes increasingly burdensome to maintain a consistent state of naming across actions.
- **Error Tracing**: Pinpointing the source of errors or unexpected behavior in the system is complex without clear identification of each callback.

## Considered Options

> 1. **Named Callbacks**: Add an optional name to the callback to make it easier to identify the source of logs. `LogAs("my-action")`.

Author: ```suggestion
1. **Named Callbacks**: Add an optional name to the callback to make it easier to identify the source of logs. `Alias("my-action")`.
```

2. **Maintain Anonymous Callbacks**: Continue using anonymous callbacks without any changes and move this responsibility to the module author. 

## Decision Outcome 

Chosen option: "Named Callbacks", because it provides a clear and consistent way to identify the source of logs and makes the code easier to understand, debug, and test with an easy developer experience.

```ts
// Watch for changes to the UDSPackage CRD to enqueue a package for processing
When(UDSPackage)
  .IsCreatedOrUpdated()
  // Advanced CR Validation
  .Validate(validator)
  // Enqueue the package for processing
  .Reconcile(reconciler)

>   .LogAs("UDS Operator Reconciler")

Author: ```suggestion
  .Alias("UDS Operator Reconciler")
```

```

## Benefits

- Simpler developer experience by reducing workload on the module author.
- Improved Logs and Debugging by programmatically identifying the source of logs in a repeatable fashion across modules.

## Consequences

> 1. Add another optional function to the Capability class, `LogAs`, to add a name to the callback.

Reviewer: ```suggestion
1. Add another optional function to the Capability class, `Alias`, to add a name to the callback (and add a default, generated name for Pepr to use when a dev forgets to set one).
```

Or is this too implementation focused?

Author: Agreed we should figure out a default name.
